### DISCUSSION TITLE
ExcelFunction IsMacroType questions

### DISCUSSION THREAD

**Discussion by emphasis87** [Excel-DNA/ExcelDna#684] [Q&A] [answered]
What are the (performance) implications using `[ExcelFunction(IsMacroType=true)]` - is there a reason not to always use it? 
Is there a documentation for it?
Can it be combined with IsThreadSafe?

**Comment by emphasis87** [comment]
I have found this: https://learn.microsoft.com/en-us/office/client-developer/excel/multithreaded-recalculation-in-excel

> Excel imposes an additional restriction: thread-safe functions cannot be registered as macro-sheet equivalents, and therefore cannot call XLM information functions or get the values of un-recalculated cells.

> Excel uses a single main thread to run or execute the following:
> - Microsoft Visual Basic for Applications (VBA) user-defined commands (often referred to as macros)
> - XLM macro sheet user-defined commands and functions

Not sure if the previous applies, especially since I don't know what exactly XLM is.

**Comment by keithalewis** [comment]
Excel macros. The C API calls into Excel using the [Excel4/Excel12 ](https://learn.microsoft.com/en-us/office/client-developer/excel/excel4v-excel12v) functions. The first argument is the "function number" to call. Functions numbers for functions start with `xlf` and function numbers for macros start with `xlc`. See https://xlladdins.github.io/Excel4Macros/ for documentation of the arguments used for macros.

**Comment by govert** [comment]
Functions that are _both_ marked as `IsMacroType=true` and have at least one parameter marked as `AllowReference=true` will be considered volatile by Excel (even if marked `IsVolatile=false`). You can internally clear the flag, but that might not always be a good idea.

I suspect functions registered as `IsMacroType=true` might be considered differently in the dependency processing, though I have no evidence or documentation to offer.

**Comment by keithalewis** [comment, marked as answer]
I'm not sure what `AllowReference=true` means in Excel-DNA. In the Excel C API pointers to XLOPER arguments can be specified as either type "Q" or type "U". If type "U" then Excel can hand you a reference and the function becomes volatile. If type "Q" then Excel will handily coerce that to the value that the reference points to and will not be marked volatile. 

I believe `IsMacroType=true` allows functions to fetch cell values that are stale and call information functions that do not change the state of the workbook. It does not make a function volatile.
